File: python_code/variable_selection/c4_stepwise_lm_model.py
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import KFold
import statsmodels.api as sm
import os


def c4_stepwise_lm_model(Pred, Resp):
    columnNames = ['Model', 'R2', 'adj-R2', 'AIC', 'R2CV']
    dataTable = pd.DataFrame(columns=columnNames)

    X = sm.add_constant(Pred)
    mdl_lm = sm.OLS(Resp, X).fit()

    kf = KFold(n_splits=5, shuffle=True, random_state=1)
    yPredCV = np.zeros(len(Resp))
    for train_idx, test_idx in kf.split(Pred):
        X_train, X_test = Pred[train_idx], Pred[test_idx]
        y_train = Resp[train_idx]
        mdl_cv = LinearRegression().fit(X_train, y_train)
        yPredCV[test_idx] = mdl_cv.predict(X_test)

    SSE = np.sum((yPredCV - Resp)**2)
    SST = np.sum((np.mean(Resp) - Resp)**2)
    R2cv = 1 - (SSE / SST)

    data = {'Model': 'linear', 'R2': mdl_lm.rsquared, 'adj-R2': mdl_lm.rsquared_adj,
            'AIC': mdl_lm.aic, 'R2CV': R2cv}
    dataTable = pd.concat([dataTable, pd.DataFrame([data])], ignore_index=True)

    base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    save_dir = os.path.join(base_dir, 'saved_models')
    os.makedirs(save_dir, exist_ok=True)
    np.savez(os.path.join(save_dir, 'stepwiselm_model.npz'),
             model=mdl_lm, perf=dataTable.to_records())

    return mdl_lm, dataTable

File: variable_selection/c4_stepwise_lm_model.m
function c4_stepwise_lm_model(Pred, Resp)

Pred=array2table(Pred);
Resp=array2table(Resp);

    [perf,model]=mdlscreen(Pred,Resp);

save stepwiselm_model.mat model perf
disp('');
end



function [dataTable,megaModel]= mdlscreen(Pred,Yi)
columnNames = {'Model', 'DF', 'SumSqr','F-value','p-value','R2','adj-R2','AIC','AICc','R2CV'};
dataTable = table('Size', [0, numel(columnNames)], 'VariableNames', columnNames,...
    'VariableTypes',{'string','double','double','double','double','double','double','double','double','double'});
megaModel=cell(1,1);
    mdl_lm=stepwiselm(Pred{:,:},Yi{:,:},'constant','Upper','linear');
    rng(1);
    cv = cvpartition(height(Pred), 'KFold', 5);
    yPredCV = zeros(height(Pred), 1);
    for k = 1:cv.NumTestSets
        trainIdx = training(cv,k);
        testIdx = test(cv,k);
        mdl_cv = fitlm(Pred(trainIdx,:), Yi(trainIdx,:), 'constant','Upper','linear');
        yPredCV(testIdx) = predict(mdl_cv, Pred(testIdx, :));
    end
    SSE=sum((yPredCV-Yi).^2);
    SST=sum((mean(Yi)-Yi).^2);
    R2cv=1-(SSE/SST);
    % for k = 1:cv.NumTestSets
    %     trainIdx = training(cv, k);
    %     testIdx = test(cv, k);
    %     mdl_cv = stepwiselm(Pred(trainIdx, :), Yi{trainIdx, :},'constant','Upper','linear');
    %     yPredCV(testIdx) = predict(mdl_cv, Pred(testIdx, :));
    % end
    % R2cv = corr(yPredCV, Yi{:,:})^2;
megaModel(1)={mdl_lm};
tab=anova(mdl_lm,'summary');
data={{'linear'},tab.DF(2),tab.SumSq(2),tab.F(2),tab.pValue(2),mdl_lm.Rsquared.Ordinary,mdl_lm.Rsquared.Adjusted,mdl_lm.ModelCriterion.AIC,mdl_lm.ModelCriterion.AICc,R2cv};
dataTable(1,:)=cell2table(data, 'VariableNames', columnNames);
disp(dataTable)
disp('');
end

